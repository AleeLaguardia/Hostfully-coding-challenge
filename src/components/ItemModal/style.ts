import styled from "styled-components";

export const Container = styled.div`
  display: grid;
  grid-template-columns: calc(30% - 5px) calc(70% - 5px);
  grid-template-rows: 1fr 1fr;
  grid-gap: 10px;
  padding: 10px;
  height: 600px;
  width: 700px;
  background-color: ${(props) => props.theme.colors.alabaster};
  border-radius: 15px;

  @media (max-width: 740px) {
    grid-template-columns: calc(40% - 5px) calc(60% - 5px);
  }
`;

export const ImageContainer = styled.div`
  display: flex;
  height: 100%;
  width: 100%;
  border-radius: 10px;

  img {
    height: 100%;
    width: 100%;
    border-radius: 10px;
    object-fit: cover;
  }
`;

export const MapContainer = styled.div`
  display: flex;
  height: 100%;
  width: 100%;
  border-radius: 10px;
`;

export const InfoContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  height: 100%;
  width: 100%;
  border-radius: 10px;

  .title {
    font-family: ${(props) => props.theme.fonts.poppins};
    font-weight: 500;
    color: ${(props) => props.theme.colors.purple};
    font-size: ${(props) => props.theme.sizes.xxxl};

    @media (max-width: 700px) {
      font-size: ${(props) => props.theme.sizes.l};
    }
  }

  .address {
    font-family: ${(props) => props.theme.fonts.poppins};
    font-weight: 500;
    color: ${(props) => props.theme.colors.lightPurple};
    font-size: ${(props) => props.theme.sizes.s};

    @media (max-width: 700px) {
      font-size: ${(props) => props.theme.sizes.xs};
    }
  }

  .description {
    width: 70%;
    margin-top: 20px;
    font-family: ${(props) => props.theme.fonts.poppins};
    font-weight: 500;
    color: ${(props) => props.theme.colors.emperor};
    font-size: ${(props) => props.theme.sizes.xs};
    text-align: center;

    @media (max-width: 700px) {
      font-size: ${(props) => props.theme.sizes.xxxs};
      margin-top: 10px;
      overflow: hidden;
      display: -webkit-box;
      -webkit-box-orient: vertical;
      -webkit-line-clamp: 5;
      text-overflow: ellipsis;
      width: 100%;
    }
  }

  .price-container {
    border: 1px solid ${(props) => props.theme.colors.lynch};
    border-radius: 10px;
    display: flex;
    width: 100%;
    justify-content: space-between;
    padding: 0 10px;
    margin-top: auto;
    padding: 10px;

    span {
      font-family: ${(props) => props.theme.fonts.poppins};
      font-weight: 500;
      color: ${(props) => props.theme.colors.lynch};
      font-size: ${(props) => props.theme.sizes.xs};
    }
  }
`;

export const BookingContainer = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 100%;
  border-radius: 10px;
  align-items: center;

  .input-container {
    display: flex;
    height: 100%;
    width: 100%;
    flex-direction: row;
    gap: 10px;
  }
`;

export const BookingTitle = styled.div`
  display: flex;
  width: fit-content;
  height: fit-content;

  span {
    font-family: ${(props) => props.theme.fonts.poppins};
    font-weight: 500;
    color: ${(props) => props.theme.colors.purple};
    font-size: ${(props) => props.theme.sizes.m};
  }
`;

export const UserInfoContainer = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 100%;
  justify-content: space-between;

  .dropdown-container {
    position: relative;
    display: flex;
    margin-bottom: 10px;
  }
`;

export const ButtonContainer = styled.div`
  display: flex;
  flex-direction: row;
  height: fit-content;
  width: 100%;
  justify-content: flex-end;
  align-items: flex-end;
  gap: 10px;
  margin-top: 5px;


  @media (max-width: 700px) {
    margin-top: 0;
  }
`;