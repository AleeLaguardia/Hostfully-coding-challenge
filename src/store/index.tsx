import { configureStore } from "@reduxjs/toolkit";
import userReducer from './slice/userInfoSlice';
import reservationReducer from "./slice/reservationSlice";

const store = configureStore({
  reducer: {
    user: userReducer,
    reservation: reservationReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;