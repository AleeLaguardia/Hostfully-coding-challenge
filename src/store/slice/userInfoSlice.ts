import { createSlice, PayloadAction } from "@reduxjs/toolkit";

interface UserState {
  destination: string;
  adults: string;
  children: string;
  date?: any;
};

const initialState: UserState = {
  destination: '',
  adults: '',
  children: '',
  date: [new Date(), new Date(new Date().getTime() + 24 * 60 * 60 * 1000)],
};

const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    setUser: (state, action: PayloadAction<UserState>) => {
      state.destination = action.payload.destination;
      state.adults = action.payload.adults;
      state.children = action.payload.children;
      state.date = action.payload.date;
    },
    updateUser: (state, action: PayloadAction<Partial<UserState>>) => {
      return { ...state, ...action.payload };
    },
  },
});

export const { setUser, updateUser } = userSlice.actions;
export default userSlice.reducer;