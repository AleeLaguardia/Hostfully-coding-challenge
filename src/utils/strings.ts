export const PATH = {
  HOME: '/',
  BOOKING: '/booking',
  RESERVATION: '/reservation',
};

export const HOME = {
  TITLE_FORM: "Book your stay right now!",
  TITLE_PIC: "Book Your Stay",
  ADULTS: "How many adults",
  CHILDREN: "How many children",
}

export const BOOKING = {
  NO_DATA: "No hotels found in that area",
  ALSO_TRY: "But you can also try:"
};