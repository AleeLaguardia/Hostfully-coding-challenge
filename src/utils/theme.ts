import { DefaultTheme } from "styled-components";

import {
  white,
  alabaster,
  emperor,
  galery,
  gray,
  black,
  purple,
  purpleHover,
  darkPurple,
  radicalRed,
  lightPurple,
  silverChalice,
  alto,
  lynch,
  selectiveYellow,
  chelseaGem,
} from './styling/colors';
import {
  poppins,
  playwrite,
} from './styling/fonts';
import {
  xs,
  xxs,
  xxxs,
  s,
  m,
  l,
  xl,
  xxxl,
} from './styling/sizes';

export const theme: DefaultTheme = {
  colors: {
    white,
    alabaster,
    emperor,
    galery,
    gray,
    black,
    purple,
    purpleHover,
    darkPurple,
    radicalRed,
    lightPurple,
    silverChalice,
    alto,
    lynch,
    selectiveYellow,
    chelseaGem,
  },
  fonts: {
    poppins,
    playwrite,
  },
  sizes: {
    xs,
    xxs,
    xxxs,
    s,
    m,
    l,
    xl,
    xxxl,
  },
}