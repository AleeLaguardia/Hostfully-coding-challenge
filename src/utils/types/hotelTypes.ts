import { LatLngExpression } from "leaflet";

export type Location = {
  coordinates: LatLngExpression;
  type: string;
}

export type Room = {
  BaseRate: number;
  BedOptions?: string;
  Description?: string;
  Description_fr?: string;
  SleepsCount?: number
  SmokingAllowed?: boolean;
  Tags?: string[];
  Type?: string;
}

export type AddressType = {
  StreetAddress: string;
  City: string;
  StateProvince: string;
  PostalCode: string;
  Country: string;
}

export type Hotel = {
  HotelId: string;
  ImageSource: string;
  HotelName: string;
  Description: string;
  Description_fr: string;
  Category: string;
  Tags: string[];
  ParkingIncluded: boolean,
  LastRenovationDate: string;
  Rating: number;
  Address: AddressType;
  Rooms: Room[];
  Location: Location;
}