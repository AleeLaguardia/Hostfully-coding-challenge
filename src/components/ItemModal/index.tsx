import React, { useState } from "react";
import { Container, ImageContainer, MapContainer, InfoContainer, BookingContainer, UserInfoContainer, BookingTitle, ButtonContainer } from "./style";
import { Hotel } from "../../utils/types/hotelTypes";
import MapComponent from "../MapComponent";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../store";
import { format } from "date-fns";
import Input from "../Input";
import Dropdown from "../Dropdown";
import { theme } from "../../utils/theme";
import Button from "../Button";
import { addReservation } from "../../store/slice/reservationSlice";

interface Props {
  hotel: Hotel;
  ref: React.RefObject<HTMLDivElement> | null;
  totalPrice: string;
  isModalOpen: boolean;
  setIsModalOpen: React.Dispatch<React.SetStateAction<boolean>>;
}

const options = ['Credit Card', 'Debit Card', 'Money', 'Gift Card'];

const ItemModal: React.FC<Props> = ({ hotel, ref, totalPrice, isModalOpen, setIsModalOpen }) => {
  const [isDropdownOpen, setIsDropdownOpen] = useState<boolean>(false);
  const [option, setOption] = useState<string>('');

  const [name, setName] = useState<string>('');
  const [phone, setPhone] = useState<string>('');
  const [email, setEmail] = useState<string>('');

  const [nameError, setNameError] = useState<boolean>(false);
  const [phoneError, setPhoneError] = useState<boolean>(false);
  const [emailError, setEmailError] = useState<boolean>(false);

  const dispatch = useDispatch();
  const reservation = useSelector((state: RootState) => state.reservation);

  const { date } = useSelector((state: RootState) => state.user);

  const { StreetAddress, City, StateProvince } = hotel.Address;

  const handleConfirm = () => {
    validateExistingBooking();
  };

  const validateEmail = (email: string): boolean => {
    const regex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    const isValid = regex.test(email);

    if (isValid) {
      setEmailError(false);
    } else {
      setEmailError(true);
    }

    return isValid;
  };

  const validateExistingBooking = () => {
    const dispatchReservation = () => {
      if (name.length > 0 && phone.length > 0 && email.length > 0) {
        const isEmailValid = validateEmail(email);

        if (isEmailValid) {
          dispatch(addReservation({
            name,
            phone,
            email,
            paymentMethod: option.length > 0 ? option : options[0],
            date,
            hotel,
          }));

          setIsModalOpen(false);
        }
      } else {
        if (name.length === 0) {
          setNameError(true);
        } else {
          setNameError(false);
        }

        if (phone.length === 0) {
          setPhoneError(true);
        } else {
          setPhoneError(false);
        }

        if (email.length === 0) {
          setEmailError(true);
        } else {
          setEmailError(false);
        }
      }
    };

    const formatSelectedHotelDate = [format(date[0], 'MM/dd/yyyy'), format(date[1], 'MM/dd/yyyy')];

    const isExistingBooking = reservation.some((el) => {
      const formatFoundHotelDate = [format(el.date[0], 'MM/dd/yyyy'), format(el.date[1], 'MM/dd/yyyy')];
      const areDatesEqual = formatSelectedHotelDate.every((date, index) => date === formatFoundHotelDate[index]);

      return el.hotel.HotelName === hotel.HotelName && areDatesEqual;
    });

    if (isExistingBooking) {
      alert("There is already a reservation with the same date");
    } else {
      dispatchReservation();
    }
  };

  return (
    <Container onClick={(e) => e.stopPropagation()} ref={ref}>
      <ImageContainer>
        <img src={hotel.ImageSource} alt="ImageSource" />
      </ImageContainer>
      <InfoContainer>
        <span className="title">{hotel.HotelName}</span>
        <span className="address">{StreetAddress}, {City}, {StateProvince}</span>
        <span className="description">{hotel.Description}</span>
        <div className="price-container">
          <span>Checkin: {format(date[0], 'MM/dd/yyyy')}</span>
          <span>Checkout: {format(date[1], 'MM/dd/yyyy')}</span>
          <span>Price: {totalPrice}</span>
        </div>
      </InfoContainer>
      <MapContainer>
        <MapComponent position={hotel.Location.coordinates} />
      </MapContainer>
      <BookingContainer>
        <BookingTitle>
          <span>Confirm your reservation</span>
        </BookingTitle>
        <div className="input-container">
          <UserInfoContainer>
            <div className="dropdown-container">
              <Dropdown
                error={false}
                data-testid="adult-dropdown"
                placeholder="Payment Method"
                isDropdownOpen={isDropdownOpen}
                setIsDropdownOpen={setIsDropdownOpen}
                setOptionSelected={setOption}
                optionSelected={option}
                options={options}
                width="100%"
              />
            </div>
            <Input
              value={name}
              onChange={(e) => setName(e.target.value)}
              placeholder="Full name"
              error={nameError}
            />
            <Input
              value={phone}
              onChange={(e) => setPhone(e.target.value)}
              placeholder="Phone number"
              error={phoneError}
            />
            <Input
              value={email}
              onChange={(e) => setEmail(e.target.value)}
              placeholder="Email"
              error={emailError}
            />
            <ButtonContainer>
              <Button
                color={theme.colors.white}
                colorHover={theme.colors.galery}
                labelColor={theme.colors.selectiveYellow}
                onClick={() => setIsModalOpen(false)}
                label="Cancel"
                width="100%"
                borderColor={theme.colors.selectiveYellow}
              />
              <Button
                color={theme.colors.selectiveYellow}
                colorHover={theme.colors.chelseaGem}
                onClick={handleConfirm}
                label="Confirm"
                width="100%"
              />
            </ButtonContainer>
          </UserInfoContainer>
        </div>
      </BookingContainer>
    </Container>
  );
};

export default ItemModal;
