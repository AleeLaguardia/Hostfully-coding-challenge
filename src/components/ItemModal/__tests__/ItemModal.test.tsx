import React from 'react';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import ItemModal from '..';
import { Hotel } from '../../../utils/types/hotelTypes';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { theme } from '../../../utils/theme';
import { Provider } from 'react-redux';
import store from '../../../store';
import * as reservationSlice from '../../../store/slice/reservationSlice';

jest.mock('../../MapComponent', () => {
  const MockMapComponent: React.FC<{ position: [number, number] }> = ({ position }) => (
    <div data-testid="mock-map-component">
      {/* Simulação de conteúdo do MapComponent */}
      <span>Mocked Map</span>
      <span>Latitude: {position[0]}</span>
      <span>Longitude: {position[1]}</span>
    </div>
  );
  return MockMapComponent;
});

describe('ItemModal Component', () => {
  const mockHotel: Hotel = {
    HotelId: '1',
    HotelName: 'Test Hotel',
    Description: 'This is a test description for the test hotel.',
    ImageSource: 'https://example.com/image.jpg',
    Address: {
      StreetAddress: '123 Test St',
      City: 'Test City',
      StateProvince: 'TS',
      Country: 'USA',
      PostalCode: '000000',
    },
    Category: 'Test Category',
    Location: {
      coordinates: [1, 1],
      type: 'test type'
    },
    Description_fr: 'Description test',
    LastRenovationDate: 'Renovation Test',
    ParkingIncluded: false,
    Rating: 5.0,
    Tags: [],
    Rooms: [
      {
        BaseRate: 100,
      },
    ],
  };

  const renderWithProviders = (ui: React.ReactElement) => {
    return render(
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <Provider store={store}>
            {ui}
          </Provider>
        </ThemeProvider>
      </BrowserRouter>
    )
  }

  it('renders with default styling', () => {
    const { asFragment } = renderWithProviders(
      <ItemModal
        hotel={mockHotel}
        ref={null}
        totalPrice="$200"
        isModalOpen={true}
        setIsModalOpen={jest.fn()}
      />
    );

    expect(asFragment).toMatchSnapshot();
  })

  it('handles confirmation with valid input', async () => {
    const mockDispatch = jest.fn();
    const mockAddReservation = jest.spyOn(reservationSlice, 'addReservation');

    jest.mock('react-redux', () => ({
      ...jest.requireActual('react-redux'),
      useDispatch: () => mockDispatch,
      useSelector: () => ({
        reservation: [],
        user: { date: [new Date(), new Date()] },
      }),
    }));

    renderWithProviders(
      <ItemModal
        hotel={mockHotel}
        ref={null}
        totalPrice="$200"
        isModalOpen={true}
        setIsModalOpen={jest.fn()}
      />
    );

    fireEvent.change(screen.getByPlaceholderText('Full name'), { target: { value: 'John Doe' } });
    fireEvent.change(screen.getByPlaceholderText('Phone number'), { target: { value: '1234567890' } });
    fireEvent.change(screen.getByPlaceholderText('Email'), { target: { value: 'john.doe@example.com' } });

    expect(screen.getByPlaceholderText('Full name')).toHaveValue('John Doe');
    expect(screen.getByPlaceholderText('Phone number')).toHaveValue('1234567890');
    expect(screen.getByPlaceholderText('Email')).toHaveValue('john.doe@example.com');
  });
});
