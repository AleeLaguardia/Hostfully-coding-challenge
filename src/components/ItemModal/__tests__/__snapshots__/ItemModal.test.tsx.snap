// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemModal Component renders with default styling 1`] = `[Function]`;
